engine/crates/engine-v2/src/sources/graphql/request/prepare.rs

```diff
@@ -196,23 +258,93 @@ impl QueryBuilderContext {
             .fields_ordered_by_parent_entity_id_then_position()
             .into_iter()
             .chunk_by(|field| field.definition().parent_entity_id);
+
         for (entity_id, fields) in entity_to_fields.into_iter() {
-            if maybe_entity_id != Some(entity_id) {
-                indent_write!(
-                    buffer,
-                    "... on {} {{\n",
-                    selection_set.walker().schema().walk(entity_id).name()
-                )?;
-                buffer.indent += 1;
-            }
-            for field in fields {
-                self.write_field(buffer, field)?;
-            }
-            if maybe_entity_id != Some(entity_id) {
-                buffer.indent -= 1;
-                indent_write!(buffer, "}}\n")?;
+            let fields = fields.collect_vec();
+            let entity = selection_set.walker().schema().walk(entity_id);
+            let in_same_entity = maybe_entity_id == Some(entity_id);
+
+            let rendering = SelectionSetRendering::new(
+                maybe_entity_id,
+                maybe_selection_set_type,
+                entity,
+                self.subgraph_id,
+                selection_set,
+            );
+
+            match rendering {
+                SelectionSetRendering::InterfaceWithPartialFragment(interface, subgraph_id) => {
+                    for r#type in interface.possible_types_ordered_by_typename() {
+                        if !r#type.subgraph_implements_interface(&subgraph_id, &interface.id()) {
+                            continue;
```

Comment:
I don't understand this part — do we skip rendering the inline fragment in the subgraph request if we can't be sure the subgraph has all the implementers of the interface?

Reply:
So what happens here is we have something like a union in the return type, or interface. Let's say this is `Product`, which is a union.

We do a query:

```graphql
query {
  products {
    ... on Node {
      id
    }
  }
}
```

This now went to subgraph B as-is. Subgraph B has ovens and toasters, but only toaster implements `Node` in B. Both, the ovens and toasters are part of the `Product` union.

This is our federated query:

```graphql
type Query
  @join__type(graph: A)
  @join__type(graph: B)
{
  products: [Product] @join__field(graph: A)
}
```

And in subgraph A this looks like:

```graphql
type Query {
  products: [Product]
}
```

So we try to query with an interface fragment. But the interface is implemented only in federated level, not in all subgraphs. The subgraph A returns products, which can be of oven or of toaster. If we send down to the subgraph the query as-is (as we do in main right now):

```graphql
query {
  products {
    ... on Node {
      id
    }
  }
}
```

This would return an error because the query returns ovens and toasters, but oven does not implement node in A:

```json
{
  "data": {
    "products": [
      null,
      null,
      {
        "id": "toaster1"
      },
      {
        "id": "toaster2"
      }
    ]
  },
  "errors": [
    {
      "message": "Error decoding response from upstream: Missing required field named 'id' at line 1 column 23",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "products",
        0
      ],
      "extensions": {
        "code": "SUBGRAPH_INVALID_RESPONSE_ERROR"
      }
    },
    {
      "message": "Error decoding response from upstream: Missing required field named 'id' at line 1 column 26",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "products",
        1
      ],
      "extensions": {
        "code": "SUBGRAPH_INVALID_RESPONSE_ERROR"
      }
    }
  ]
}
```

Now, what we do in these first three matches is we take a look into the query fragment, and then we check this subgraph does it implement all possible instances of the interface. If not, we change the query to explicitly expand into all the fragments, so the correct query to the subgraph becomes:

```graphql
query {
  products {
    ... on Oven {
      id
    }
    ... on Toaster {
      id
    }
  }
}
```

```json
{
  "data": {
    "products": [
      {
        "id": "oven1"
      },
      {
        "id": "oven2"
      },
      {
        "id": "toaster1"
      },
      {
        "id": "toaster2"
      }
    ]
  }
}
```

Which returns the correct data. This is now repeated in the first three matches so that we never try to send a fragment to this subgraph that it cannot represent.

Reply:
I think that's false at this point we know already that not every object in this subgraph implements the interface. But we don't want to skip them. you must filter on object that are in said subgraph. So:

```rust
                   for object in interface.possible_types_ordered_by_typename() {
                        if !object.only_resolvable_in.contains(subgraphs) && !object.only_resolvable_in.is_empty() {
                            continue;
                        }
```

Reply:
However, you can use `subgraph_implements_interface` to create fewer fragments. You only need to add object type conditions for those that don't implement the interface so:
```rust
add_interface_fragment = false
for object in interface.possible_types_ordered_by_typename() { 
   if !object.only_resolvable_in.contains(subgraphs) && !object.only_resolvable_in.is_empty() { continue; }
   if !r#type.subgraph_implements_interface(&subgraph_id, &interface.id()) {
     write_entity_fields(...) // for object
   } else {
add_interface_fragment = true
}
   
}

if add_interface_fragment {
     write_entity_fields(...) // for object
}
```